ambry-mysql/src/main/java/com/github/ambry/mysql/BatchUpdater.java

```diff
@@ -0,0 +1,158 @@
+ *           statement.setString(1, name);
+ *       });
+ *   }
+ *   batch.flush();
+ * </pre>
+ */
+public class BatchUpdater {
+  private static final Logger logger = LoggerFactory.getLogger(BatchUpdater.class);
+  private final boolean autoCommit;
+  private final PreparedStatement statement;
+  private final int maxBatchSize;
+  private final MySqlDataAccessor dataAccessor;
+  private final String tableName;
+  private int currentBatchSize = 0;
+  private int committedBatchSize = 0;
+  // The time when the first statement is added
+  private long startTime = 0;
+
+  /**
+   * Constructor to instantiate a {@link BatchUpdater}.
+   * @param dataAccessor The {@link MySqlDataAccessor}.
+   * @param sql The sql command to prepare the statement.
+   * @param tableName The name of the table that update statements are targeting at.
+   * @param maxBatchSize The max batch size.
+   * @throws SQLException
+   */
+  public BatchUpdater(MySqlDataAccessor dataAccessor, String sql, String tableName, int maxBatchSize)
+      throws SQLException {
+    this.dataAccessor = Objects.requireNonNull(dataAccessor, "MySqlDataAccessor is empty");
+    this.tableName = Objects.requireNonNull(tableName, "TableName is empty");
+    if (maxBatchSize < 0) {
+      throw new IllegalArgumentException("MaxBatchSize is not valid: " + maxBatchSize);
+    }
+    this.maxBatchSize = maxBatchSize;
+    try {
+      autoCommit = dataAccessor.getAutoCommmit();
+      dataAccessor.setAutoCommit(false);
+      statement = dataAccessor.getPreparedStatement(sql, true);
+      statement.clearBatch();
+    } catch (SQLException e) {
+      dataAccessor.onException(e, BatchUpdate);
+      logger.error("Failed to prepare for batch insert on {}", tableName, e);
+      throw e;
+    }
+  }
+
+  /**
+   * Add statement to the batch. Since the sql statement provided to the constructor is a prepared statement,
+   * this method only takes a lambda to supply values to the {@link PreparedStatement}.
+   * @param valueSupplier The consumer to supply the values. The {@link PreparedStatement} is the parameter to this
+   *                      consumer. It can also throws an {@link SQLException}.
+   * @throws SQLException
+   */
+  protected synchronized void addUpdateToBatch(GenericThrowableConsumer<PreparedStatement, SQLException> valueSupplier)
+      throws SQLException {
+    boolean failed = false;
+    try {
+      if (startTime == 0) {
+        startTime = System.currentTimeMillis();
+      }
+      if (maxBatchSize != 0 && currentBatchSize >= maxBatchSize) {
+        executeBatchAndCommit();
+      }
+      valueSupplier.accept(statement);
+      statement.addBatch();
+      currentBatchSize++;
+    } catch (SQLException e) {
+      failed = true;
+      dataAccessor.onException(e, BatchUpdate);
+      logger.error("Failed to add batch on {}, already committed {}, rolling back", tableName, committedBatchSize, e);
+      dataAccessor.rollback();
```

Comment:
Minor: Just thinking if it is safer to call dataAccessor.rollback() before calling dataAccessor.onException() since the latter method has code to close the db connection in case the SQL Exception is a transient one.

Reply:
I don't think this is minor.  The rollback method will throw IllegalStateException if connection has been closed.  Do we have a unit test to simulate this scenario?

Reply:
I checked the codecov, and the exception clauses are red (not covered).  I recommend adding a BatchUpdater test class or adding test cases to the stats store test to trigger the exceptions.

Reply:
Unit test added.